### DISCUSSION TITLE
Cache first policy

### DISCUSSION THREAD

**Discussion by peterplamenovpenchev** [coil-kt/coil#1175] [Ideas]
Hi,

I am using the `1.4.0` version of the library & it's been a blast.
There is one thing that appears to be missing though: cache first policy when loading images.

Here is the setup that I am working with:
1. Have a basic client (Android) side setup of the Coil library
2. Load images from a server which returns the `Last-Modified` header
3. This `Last-Modified` header is used in subsequent requests as `If-Modified-Since`
4. The subsequent requests return either:
- `304` -> Not modified
- `200 + body` -> if the image was modified since the last request

This as you know is pretty standard setup so nothing interesting by far.

Now what I think would be cool is to have a cache first policy. I am thinking of the following behaviour:
1. Trigger an image request
2. Return the image from the cache (if it was cached)
3. In the background - make an HTTP request for the image
4. Update the image if a new one is returned from the network for that URL

Sadly I haven't got the chance to play around with the `2.0.0` version - does it support that kind of behaviour?
And if not - is there a chance we would see something like that in the future?

Thanks!

**Comment by colinrtwhite** [comment]
Unfortunately I don't think this is possible to implement since Coil's image pipeline is built around returning one drawable per request. `HttpFetcher` can only return one `ImageSource` to be decoded.

**Reply by serhii-petrenko-dev** [reply]
I'm curious if it's possible to make coil:
1. Return the image from the cache, if cached
2. If not - do a network request

Usecase for this described in [question](https://stackoverflow.com/questions/71141940/coil-image-caching):
`I load my images into an android app with the Coil library. And it instantly shows me cached images if there is no internet. But when there is an internet connection, the Coil loads images again, and for a while I see the placeholder. How can I make it show me the cached images instantly even if there is an internet connection?`

**Reply by dokar3** [reply]
@serhii-petrenko-dev Maybe you are facing a cache headers problem? Disable the support for cache headers may help:
```kotlin
val imageLoader = ImageLoader.Builder(context)
        .respectCacheHeaders(false)
        .build()
Coil.setImageLoader(imageLoader)
```

**Reply by serhii-petrenko-dev** [reply]
@dokar3, you are right. After disabling cache headers image display lag disappeared! Thanks!
